src/credentials/GroupDataProviderImpl.h

```diff
@@ -72,95 +79,83 @@ class GroupDataProviderImpl : public GroupDataProvider
     {
     public:
-        AllStatesIterator(GroupDataProviderImpl & provider);
+        GroupKeyIteratorImpl(GroupDataProviderImpl & provider, chip::FabricIndex fabric_index);
         size_t Count() override;
-        bool Next(GroupState & item) override;
+        bool Next(GroupKey & output) override;
         void Release() override;
 
     private:
         GroupDataProviderImpl & mProvider;
-        uint16_t mIndex    = 0;
-        size_t mCount      = 0;
-        size_t mTotalCount = 0;
+        chip::FabricIndex mFabric = kUndefinedFabricIndex;
+        uint16_t mNextId          = 0;
+        size_t mCount             = 0;
+        size_t mTotal             = 0;
     };
 
-    class FabricStatesIterator : public GroupStateIterator
+    class EndpointIteratorImpl : public EndpointIterator
     {
     public:
-        FabricStatesIterator(GroupDataProviderImpl & provider, chip::FabricIndex fabric_index);
+        EndpointIteratorImpl(GroupDataProviderImpl & provider, chip::FabricIndex fabric_index);
         size_t Count() override;
-        bool Next(GroupState & item) override;
+        bool Next(GroupEndpoint & output) override;
         void Release() override;
 
     private:
         GroupDataProviderImpl & mProvider;
         chip::FabricIndex mFabric = kUndefinedFabricIndex;
-        uint16_t mIndex           = 0;
-        size_t mCount             = 0;
-        size_t mTotalCount        = 0;
+        chip::GroupId mFirstGroup = kUndefinedGroupId;
+        uint16_t mGroup           = 0;
+        size_t mGroupIndex        = 0;
+        size_t mGroupCount        = 0;
+        uint16_t mEndpoint        = 0;
+        size_t mEndpointIndex     = 0;
+        size_t mEndpointCount     = 0;
     };
 
     class KeySetIteratorImpl : public KeySetIterator
     {
     public:
         KeySetIteratorImpl(GroupDataProviderImpl & provider, chip::FabricIndex fabric_index);
         size_t Count() override;
-        bool Next(KeySet & item) override;
+        bool Next(KeySet & output) override;
         void Release() override;
 
     private:
         GroupDataProviderImpl & mProvider;
         chip::FabricIndex mFabric = kUndefinedFabricIndex;
         uint16_t mNextId          = 0;
         size_t mCount             = 0;
-        size_t mIndex             = 0;
+        size_t mTotal             = 0;
     };
 
     chip::PersistentStorageDelegate & mStorage;
     bool mInitialized = false;
-    BitMapObjectPool<AllGroupMappingsIteratorImpl, kIteratorsMax> mAllGroupsIterators;
-    BitMapObjectPool<GroupMappingIteratorImpl, kIteratorsMax> mEndpointGroupsIterators;
-    BitMapObjectPool<AllStatesIterator, kIteratorsMax> mAllStatesIterators;
-    BitMapObjectPool<FabricStatesIterator, kIteratorsMax> mFabricStatesIterators;
-    BitMapObjectPool<KeySetIteratorImpl, kIteratorsMax> mKeySetIterators;
+    BitMapObjectPool<GroupInfoIteratorImpl, kIteratorsMax, OnObjectPoolDestruction::IgnoreUnsafeDoNotUseInNewCode>
```

Comment:
Don't do this; call `BitMapObjectPool::ReleaseAll()` in the enclosing class shutdown or destructor if there's no other management of the objects.

Reply:
ReleaseAll() is now called from the Finish() function.